lib/withOnyx.d.ts

```diff
@@ -0,0 +1,23 @@
+import {IsEqual} from 'type-fest';
+import {Key, CollectionKey, Selector, Value} from './types';
+
+type Mapping<TComponentProps, TOnyxProps, TMappingKey extends keyof TOnyxProps, TOnyxKey extends Key | CollectionKey> = IsEqual<Value[TOnyxKey] | null, TOnyxProps[TMappingKey]> extends true
+    ? {
+          key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
+          canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+          initWithStoredValues?: boolean;
+          selector?: Selector<TOnyxKey>;
+      }
+    : never;
+
+type KeyValueMapping<TComponentProps, TOnyxProps, TMappingKey extends keyof TOnyxProps> = {
+    [TOnyxKey in Key | CollectionKey]: Mapping<TComponentProps, TOnyxProps, TMappingKey, TOnyxKey>;
```

Comment:
is this going to produce something like this?

```ts
{
  onyxKey1: never;
  onyxKey2: never;
  onyxKey3: never;
  ...
};
```

Reply:
ah I see it now!!!

`KeyValueMapping` is going to produce something like this `never | never | { key: ...; canEvict... } | never | never...`

so this is essentially `{ key: ...; canEvict... }`

Nice idea to extract the onyx key!!!

Reply:
wait but this clever code is using value type to check if the two keys are the same.

but if there are two keys in Onyx that hold the same value type (for example, `string`), this is going to break? `KeyValueMapping` will be a union type of bunch of `Mapping` if my understanding is correct?

In the usage pattern of Onyx in App, it's unlikely two keys have data of the same data type, but still the typing doesn't reflect the API?

Reply:
Sorry for that @hayata-suenaga , I think the name I used for the type is causing confusion. I can explain better tomorrow and will change some names to reflect better its meaning 😅

Reply:
This is more contracted example code that illustrates the issue I provided above. Note that even though I passed `key2` to `key`, TS doesn't complain

```ts
import { IsEqual } from "type-fest";

/* eslint-disable @typescript-eslint/no-unused-vars */
type Key = "key1" | "key2" | "key3";

type Value = {
  key1: string;
  key2: string;
  key3: boolean;
};

type Mapping<
  TProps,
  TPropName extends keyof TProps,
  TKey extends Key
> = IsEqual<Value[TKey], TProps[TPropName]> extends true
  ? {
      key: TKey;
      value: Value[TKey];
    }
  : never;

type Helper<TProps, TPropName extends keyof TProps> = {
  [K in Key]: Mapping<TProps, TPropName, K>;
}[Key];

function testFn<TProps>(mapping: {
  [K in keyof TProps]: Helper<TProps, K>;
}) {
  console.log(mapping);
}

type Props = {
  prop1: string;
  prop2: string;
};

testFn<Props>({
  prop1: {
    key: "key1",
    value: "test",
  },
  prop2: {
    key: "key1",
    value: "test",
  },
});
```

Reply:
I guess this is not an issue though? We don't have any information that ties prop name to particular Onyx key... so it's impossible to enforce anything about that.

Reply:
wait! but the type of prop value is not always same as the value for onyx

because of the selector...

Reply:
@fabioh8010 

In the example code you provided [here](https://github.com/Expensify/react-native-onyx/pull/267#issuecomment-1632060418),

```ts
type OnyxProps = Nullable<{
    account: Account;
    locale: string;
    lastOpenedPublicRoomId: number;
}>;
```

The type of `account` should be `string` as we're selecting `name` from `Account` using the `selector`.

This is going to cause a problem I suggested in the [above comment](https://github.com/Expensify/react-native-onyx/pull/267#discussion_r1263092472)

Reply:
@hayata-suenaga I'm checking this 👍